src/funnels/Depositor.sol

```diff
@@ -0,0 +1,293 @@
+        else revert("Depositor/file-unrecognized-param");
+        emit File(what, gem0, gem1, data);
+    }
+
+    struct DepositParams {
+        address gem0;
+        address gem1;
+        uint256 amt0;
+        uint256 amt1;
+        uint256 minAmt0;
+        uint256 minAmt1;
+        uint24 fee;
+        int24 tickLower;
+        int24 tickUpper;
+    }
+
+    function addLiquidity(DepositParams memory p, address to) internal returns (uint128 liquidity, uint256 amt0, uint256 amt1) {
+        bytes32 key = keccak256(abi.encode(p.gem0, p.gem1, p.fee, p.tickLower, p.tickUpper));
+        uint256 tokenId = positions[key];
+        if (tokenId == 0) {
+            PositionManagerLike.MintParams memory params = PositionManagerLike.MintParams({
+                token0: p.gem0,
+                token1: p.gem1,
+                fee: p.fee,
+                tickLower: p.tickLower,
+                tickUpper: p.tickUpper,
+                amount0Desired: p.amt0,
+                amount1Desired: p.amt1,
+                amount0Min: p.minAmt0,
+                amount1Min: p.minAmt1,
+                recipient: to,
+                deadline: block.timestamp
+            });
+            (tokenId, liquidity, amt0, amt1) = PositionManagerLike(uniV3PositionManager).mint(params);
+            positions[key] = tokenId;
+        } else {
+            PositionManagerLike.IncreaseLiquidityParams memory params = PositionManagerLike.IncreaseLiquidityParams({
+                tokenId: tokenId,
+                amount0Desired: p.amt0,
+                amount1Desired: p.amt1,
+                amount0Min: p.minAmt0,
+                amount1Min: p.minAmt1,
+                deadline: block.timestamp
+            });
+            (liquidity, amt0, amt1) = PositionManagerLike(uniV3PositionManager).increaseLiquidity(params);
+        }
+        
+        // Send leftover tokens back to buffer
+        if(amt0 < p.amt0) DepositedGemLike(p.gem0).transfer(to, p.amt0 - amt0);
+        if(amt1 < p.amt1) DepositedGemLike(p.gem1).transfer(to, p.amt1 - amt1);
+    }
+
+    function deposit(DepositParams memory p) external auth returns (uint128 liquidity, uint256 amt0, uint256 amt1) {
+        require(p.gem0 < p.gem1, "Depositor/wrong-gem-order");
+
+        require(block.timestamp >= zzz[p.gem0][p.gem1] + hops[p.gem0][p.gem1], "Depositor/too-soon");
+        zzz[p.gem0][p.gem1] = block.timestamp;
+
+        address buffer_ = buffer;
+        DepositedGemLike(p.gem0).transferFrom(buffer_, address(this), p.amt0);
+        DepositedGemLike(p.gem1).transferFrom(buffer_, address(this), p.amt1);
+        DepositedGemLike(p.gem0).approve(uniV3PositionManager, p.amt0); // TODO: cheaper to SLOAD allowance to check if we need to approve max?
+        DepositedGemLike(p.gem1).approve(uniV3PositionManager, p.amt1);
+
+        (liquidity, amt0, amt1) = addLiquidity(p, buffer_);
+        require(amt0 * amt1 <= caps[p.gem0][p.gem1], "Depositor/exceeds-cap");
```

Comment:
Need to think more about it but my current thinking is:

The caps will need to be updated when the pool liquidity changes substantially as the impact of manipulation would otherwise change correspondingly. Price changes only have an indirect effect on the pool liquidity so it seems to me that capping the amount of added liquidity is a more correct way to limit the potential for manipulation.

Reply:
Isn't the simplest way to cap the deposit amounts? I think in most cases the configurator here and the depositor that uses the automation sample on top will not know anything about liquidity terms.
IMO we should just assume that a malicious/ignorant facilitator can cause a loss of the entire `amt0` and `amt1` (same as the swapper, in which case it is `maxSrcAmt`) and not get into how likely that is for different amounts of liquidity.

Reply:
Ok I see your point. Changed.

Reply:
Ok, nice optimization with the uint128 packing.
It means that each amount can be up to 2^128 / 10**18 ~= 3.4 * 10^20.
That should be more than enough.